File: Sources/macOS/Sources/Yage/Capabilities/ShapeShifter.swift
import SwiftUI

public class ShapeShifter: Capability {
    var animationDuracy: TimeInterval = 1
    var remainingTime: TimeInterval = 0
    var targetSize: CGSize = .zero
    var delta: CGSize = .zero

    public func scaleLinearly(to size: CGSize, duracy: TimeInterval) {
        guard let subject else { return }
        targetSize = size
        delta = CGSize(
            width: size.width - subject.frame.width,
            height: size.height - subject.frame.height
        )
        animationDuracy = duracy
        remainingTime = duracy
        isEnabled = true
    }

    override public func doUpdate(with collisions: Collisions, after time: TimeInterval) {
        subject?.frame = updatedFrame(given: time)
        checkCompletion(given: time)
    }
    
    private func updatedFrame(given time: TimeInterval) -> CGRect {
        guard let subject else { return .zero }
        let delta = CGSize(
            width: time * delta.width / animationDuracy,
            height: time * delta.height / animationDuracy
        )
        return CGRect(
            x: subject.frame.origin.x - delta.width / 2,
            y: subject.frame.origin.y - delta.height / 2,
            width: subject.frame.width + delta.width,
            height: subject.frame.height + delta.height
        )
    }

    private func checkCompletion(given time: TimeInterval) {
        remainingTime -= time
        if remainingTime <= 0.0001 {
            isEnabled = false
        }
    }
}

File: Sources/windows/yage/capabilities/shape_shifter.py
from yage.models.capability import Capability
from yage.utils.geometry import Rect, Size


class ShapeShifter(Capability):
    def __init__(self, subject):
        super().__init__(subject)
        self.animation_duracy = 1
        self.target_size = Size.zero()
        self.delta = Size.zero()
        self.remaining_time = 0

    def scale_linearly(self, size, duracy):
        self.target_size = size
        self.delta = Size(
            width=size.width - self.subject.frame.width,
            height=size.height - self.subject.frame.height
        )
        self.animation_duracy = duracy
        self.remaining_time = duracy
        self.is_enabled = True

    def do_update(self, collisions, time):
        self.subject.frame = self._updated_frame(time)
        self._check_completion(time)

    def _updated_frame(self, time):
        delta = Size(
            width=time * self.delta.width / self.animation_duracy,
            height=time * self.delta.height / self.animation_duracy
        )
        return Rect(
            x=self.subject.frame.origin.x - delta.width / 2,
            y=self.subject.frame.origin.y - delta.height / 2,
            width=self.subject.frame.width + delta.width,
            height=self.subject.frame.height + delta.height
        )

    def _check_completion(self, time):
        self.remaining_time -= time
        if self.remaining_time <= 0.0001:
            self.is_enabled = False
